### PR TITLE
i#3823 multi-phase drreg: Delay slot id label

### PR DESCRIPTION
**Pull request by abhinav92003**
Moves label that contains slot id for register spill/restore instrs to after the instr instead of
before. The free spill slot selection logic that makes use of these labels scans instrs *after*
the given one, so we may miss the label if it is placed before.

Fixes order of app val spill and tool val restore instrs after an instr that reads and writes a
spilled reg. This was to take into account the label which is now after the tool restore instr.

Adds test to verify restoration of reg that was reserved in multiple phases on a fault, for
X86 and AARCHXX.

Also adds AARCHXX variant of the multi-phase slot conflict test, and extends it to also
check proper restoration of app val (under normal operation, as opposed to under a
fault which is done by the above test). The existing test only verified whether the slot
used in different phases was different.

Adds a note to the label instrs added by drreg-test to mark instrumentation locations. This
is to avoid conflicts with other label instrs.

Issue: #3823, #2985

### CODE REVIEW

**Review by derekbruening — APPROVED** [review]

File: suite/tests/client-interface/drreg-test-shared.h [hunk 1]
```diff
@@ -66,7 +66,7 @@
  * specific tests in the client.
  * We limit to 16 bits to work on ARM.
  */
-#define DRREG_TEST_CONST(num) f1f##num
+#define DRREG_TEST_CONST(num) f1##num
```
[the remark this answers is not in the document]
Comment (abhinav92003) [reply suite/tests/client-interface/drreg-test-shared.h:69]: I add an AArch64 test that has two hex digits. So f1f##num would exceed the 16-bit immediate limit for AArch64 movs (documented in comment above).

[the remark this answers is not in the document]
Comment (derekbruening) [reply suite/tests/client-interface/drreg-test-shared.h:69]: But the existing use goes up to 13, and you added 14 -- that doesn't seem to cross a digit threshold that wasn't there before?

File: ext/drreg/drreg.c [hunk 2]
```diff
@@ -700,12 +702,23 @@ drreg_event_bb_insert_late(void *drcontext, void *tag, instrlist_t *bb, instr_t
...
-                           * are in the proper order.
-                           */
-                          restored_for_read[GPR_IDX(reg)] ? instr_get_prev(next)
-                                                          : next /*after*/);
+                /* If reads and writes, make sure tool-restore and app-spill
+                 * are in the proper order.
+                 */
+                if (restored_for_read[GPR_IDX(reg)]) {
+                    ASSERT(instr_get_prev(next) != NULL &&
+                               instr_is_label(instr_get_prev(next)) &&
```
[the remark this answers is not in the document]
Comment (abhinav92003) [reply ext/drreg/drreg.c:713]: Done.

File: suite/tests/client-interface/drreg-test-shared.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (abhinav92003) [reply suite/tests/client-interface/drreg-test-shared.h:69]: The existing two-hex-digit tests are only for X86. test14 is the first such for AArch64, which is why this didn't come up before. Other tests probably still need to be ported to AArch64.

For the same test, I'm keeping the test numbers same across all archs. That's why I added 14 directly on AArch64.

**Comment by abhinav92003** [comment]
Seems that `drreg-test` isn't one of the tests we run on ARM QEMU: https://github.com/DynamoRIO/dynamorio/blob/ed1582e1f08d998beb1d0e564df842ea49cb7550/suite/tests/CMakeLists.txt#L4806

I suppose there was some issue? I can try again.

File: suite/tests/client-interface/drreg-test-shared.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (derekbruening) [reply suite/tests/client-interface/drreg-test-shared.h:69]: (Tried out the new Conversations feature up top to see unresolved conversations -- seems they're finally adding some key features.)  OK I see.

**Comment by abhinav92003** [comment]
> Seems that `drreg-test` isn't one of the tests we run on ARM QEMU:
> 
> https://github.com/DynamoRIO/dynamorio/blob/ed1582e1f08d998beb1d0e564df842ea49cb7550/suite/tests/CMakeLists.txt#L4806
> 
> I suppose there was some issue? I can try again.

It failed. From https://github.com/DynamoRIO/dynamorio/issues/4719#issuecomment-774342094 I see that `drreg-test` was failing for ARM already. Not sure if it fails also on real ARM machine. Is it worth fixing it right away in this PR?

**Comment by derekbruening** [comment]
> It failed. From [#4719 (comment)](https://github.com/DynamoRIO/dynamorio/issues/4719#issuecomment-774342094) I see that `drreg-test` was failing for ARM already. Not sure if it fails also on real ARM machine. Is it worth fixing it right away in this PR?

I would say no.

**Comment by abhinav92003** [comment]
> > It failed. From [#4719 (comment)](https://github.com/DynamoRIO/dynamorio/issues/4719#issuecomment-774342094) I see that `drreg-test` was failing for ARM already. Not sure if it fails also on real ARM machine. Is it worth fixing it right away in this PR?
> 
> I would say no.

Okay. I'm not too worried about this PR as it doesn't introduce any ARM-only logic; even drreg barely has any. Also, the new  ARM test has same asm as one before.

**Pull request merged**